### PR TITLE
sunxi-cortexa7: add support for FriendlyElec NanoPi R1

### PR DESCRIPTION
**Pull request by goligo**
**blocked by** https://github.com/openwrt/openwrt/issues/10080

- [x] Must be flashable from vendor firmware
  - [ ] Web interface
  - [ ] TFTP
  - [x] Other: SD card image / no vendor firmware
- [x] Must support upgrade mechanism
  - [x] Must have working sysupgrade
    - [x] Must keep/forget configuration (`sysupgrade [-n]`, `firstboot`)
  - [x] Gluon profile name matches autoupdater image name
        (`lua -e 'print(require("platform_info").get_image_name())'`)
- [ ] Reset/WPS/... button must return device into config mode
- [ ] Primary MAC address should match address on device label (or packaging)
      (https://gluon.readthedocs.io/en/latest/dev/hardware.html#notes)
  - When re-adding a device that was supported by an earlier version of Gluon, a
    factory reset must be performed before checking the primary MAC address, as
    the setting from the old version is not reset otherwise.
- Wired network
  - [x] should support all network ports on the device
  - [x] must have correct port assignment (WAN/LAN)
    - On devices supplied via PoE, there is usually no explicit WAN/LAN labeling on the hardware.
      The PoE input should be the WAN port in this case.
- Wireless network (if applicable)
  - [x] Association with AP must be possible on all radios
  - [ ] Association with 802.11s mesh must work on all radios 
  - [ ] AP+mesh mode must work in parallel on all radios 
- LED mapping
  - Power/system LED
    - [x] Lit while the device is on
    - [x] Should display config mode blink sequence 
          (https://gluon.readthedocs.io/en/latest/features/configmode.html)
  - Radio LEDs - there is no radio LED
    - [ ] Should map to their respective radio
    - [ ] Should show activity
  - Switch port LEDs
    - [x] Should map to their respective port (or switch, if only one led present) 
    - [x] Should show link state and activity
- Outdoor devices only:
  - [ ] Added board name to `is_outdoor_device` function in `package/gluon-core/luasrc/usr/lib/lua/gluon/platform.lua`

### CODE REVIEW

**Comment by goligo** [comment]
- Mesh does not work - BCM43430 does not support 802.11s
- Reset button does not have any effect - neither resets nor starts config mode

**Comment by goligo** [comment]
@AiyionPrime

**Comment by rotanid** [comment]
please read the requirements, especially that devices without mesh support won't be merged without the BROKEN flag.
https://github.com/freifunk-gluon/gluon/wiki/Device-Integration

**Comment by AiyionPrime** [comment]
Did not know about the broken mesh; but having this device added as `BROKEN` would still be worthwile, as its a pretty decent offloader.

**Comment by goligo** [comment]
> please read the requirements, especially that devices without mesh support won't be merged without the BROKEN flag. https://github.com/freifunk-gluon/gluon/wiki/Device-Integration

I tend do disagree in this case, BROKEN implies there is something which can and should be fixed, which is not the case here. Like the Raspberry Pi 3, which can´t do mesh as well, but is not marked as BROKEN

**Comment by belzebub40k** [comment]
@goligo the whole target of the RPi3 is marked as broken

https://github.com/freifunk-gluon/gluon/blob/master/targets/targets.mk#L24

**Comment by AiyionPrime** [comment]
And here is the raspberrys current discussion: #1687

I think this would then be one take form the referenced PR:

> At the very least we would need to ensure that Gluon doesn't even attempt to use the built-in WLAN for meshing.

_Originally posted by @NeoRaider in https://github.com/freifunk-gluon/gluon/issues/1687#issuecomment-485495687_

**Comment by goligo** [comment]
Added BROKEN flag and removed it from supported devices list.

**Comment by blocktrron** [comment]
> Primary MAC address should match address on device label

At least the R2S does not have a unique MAC address stored anywhere on the board. I was also not able to find the MAC address extraction logic in U-Boot / OpenWrt. Any chance you have two of said boards? I suspect the MAC address is generated randomly on boot, which can lead to two devices ending up with the same MAC address (No-Go in a batman-adv network).

The R2S uses the unique MMC serial number to derive it's MAC address.

**Comment by goligo** [comment]
> which can lead to two devices ending up with the same MAC address

Yes, we experienced that painfully with early firmware on the R2S in FFMUC. I know we have one other instance of the R1 in the community, that did get a different MAC address than mine, without any manual intervention.

**Comment by rotanid** [comment]
> I tend do disagree in this case, BROKEN implies there is something which can and should be fixed, which is not the case here. Like the Raspberry Pi 3, that can´t do mesh as well, but is not marked as BROKEN

the raspberry pi 3 was integrated before we had agreed on these rules.

i wonder why we set up rules and now we tend to discuss the applicability on each occasion? (see #1986) might as well drop the rules...
The rules were set up by the maintainers, not by the public, to be able to stop discussing about each issue again and again - in the end everyone profits if there are clear rules.
everyone is able to have a Gluon fork (which even my community has, by the way)

**Comment by goligo** [comment]
> i wonder why we set up rules and now we tend to discuss the applicability on each occasion?

Rules are great and have a purpose, but as the world usually is more complex than expected, there can also be exception to rules. Rules shouldn't be followed just for the sake of following rules. However, as you might have noticed, I already have changed the PR, so there is no need to start a basic discussion about it.

File: targets/sunxi-cortexa7 [hunk 1]
```diff
@@ -15,3 +15,7 @@ device('lemaker-banana-pro', 'lemaker_bananapro', {
 device('lamobo-r1', 'lamobo_lamobo-r1', {
 	broken = true, -- AP+11s not working
 })
+
+device('friendlyarm-nanopi-r1', 'friendlyarm_nanopi-r1', {
+        broken = true, -- 11s not working
```
Comment (rotanid) [inline targets/sunxi-cortexa7:20]: sounds picky i know, but please use the same indentation as elsewhere, like 4 lines above

Comment (goligo) [reply targets/sunxi-cortexa7]: Done.

**Comment by blocktrron** [comment]
Okay, I've had a more in-depth look into U-Boot and the OpenWrt userland. Can you check if using the device with OpenWrt generates random MAC addresses on subsequent boots? In case the auto-generation is done within Linux, this is usually also echoed to the kernel ringbuffer.

In this case, a workaround similar to the one used on the R2S is necessary, as a) this platforms usually do not have a reliable entropy source and b) the WAN mac-address in Gluon would not be stable.

**Comment by goligo** [comment]
> Can you check if using the device with OpenWrt generates random MAC addresses on subsequent boots?

Yes exactly this is the case - WAN MAC address changes on every restart of the device.

**Comment by blocktrron** [comment]
@goligo Did you asses my comment regarding reading the CID of the SD card to generate a stable MAC address? I see this as the biggest blocker for this board.

**Comment by goligo** [comment]
@blocktrron That would be a patch for OpenWrt, right? Like you recently did for the R4S?

**Comment by blocktrron** [comment]
Correct, you can take it basically as a blueprint.

**Comment by goligo** [comment]
As discovered by @AiyionPrime , when trying to make progress with the stable MAC, there are more severe issues with NanoPi R1 introduced with OpenWrt 22.03 / Kernel 5.10. MMC detection is unstable and randomly assigns different device names on each boot (mmcblk0 or mmcblk2). There also seems to be an issue with networking, even when it boots, in Gluon v2022.1.1 network connection on br-wan is not functional.

**Comment by AiyionPrime** [comment]
I'll update this comment with an upstream issue reference once I opened it either today or tomorrow. This is expected to be resolvable. Hauke already suggested smth. On it.

**Comment by AiyionPrime** [comment]
First patch is online: https://patchwork.ozlabs.org/project/openwrt/patch/20221211130446.23492-1-git@aiyionpri.me/

Might have found a tester for it as well.
MMC detection is still not solved, I think.

**Comment by AiyionPrime** [comment]
Second Patch is online as well: https://patchwork.ozlabs.org/project/openwrt/patch/20221215141833.100172-1-git@aiyionpri.me/

Third is done as well, but rests until the other two patches got any answers.
Merging it without them is worthless.

**Comment by AiyionPrime** [comment]
First two patches have been merged into master.
The last one is now submitted as well.

https://patchwork.ozlabs.org/project/openwrt/patch/20221222201201.283913-1-git@aiyionpri.me/

@stintel suggested waiting for the second patches fallout and do the backport if it does not occur.

**Comment by AiyionPrime** [comment]
The third patch has been merged as well.
While doing the patches, I learnt about the `Secure ID` every H2+/H3/H5 Soc is supposed to have. Basically a 16 byte serial number.

U-Boot uses that as source for the mac-addresses and I intend to implement it the same way for the NanoPi R1.

Other than that, on OpenWrts Master the NanoPi is now fully functional, does have a MAC based on the (e)MMC and should therefore only have MAC collisions, if cheap micro SD cards are used.
(A risk, thats shared with the R2 and R4 rockchip variants as well)

So once the patches two and three have been backported and we updated our modules afterwords, this can finally continue.

**Comment by AiyionPrime** [comment]
Currently waiting on the merge of this upstream backport series:
https://patchwork.ozlabs.org/project/openwrt/list/?series=335395

**Comment by blocktrron** [comment]
Series has been backported, whats the status here?

**Comment by Djfe** [comment]
the patch series fixed the flash detection but the wifi path also changes randomly between mmc0 and mmc1. That would probably throw Gluon off as well, right?
https://github.com/openwrt/openwrt/issues/12237

**Comment by AiyionPrime** [comment]
I asked @goligo in early april whether he wants to put more effor tinto this right now, but did not receive an answer yet.
For me the priority of this device falls behind nearly every other thing sitting on my desk rn.

The current next steps would be:
- [ ] verify my upstream patches not only allow the device to boot in 22.03 and master
- [ ] but also have a mac-address depending on the hardware

The pitfall is that this device comes in two variants: with eMMC and without.

@goligo and me have access to a variant each, but apparently lack the time to dig into this further right now.

@goligo I hope you don't mind me closing this for the moment.
Let me know if you intend to pursue this and I'll reopen it at once.

**Pull request closed without merging**